### PR TITLE
Condition handling (for issue #142)

### PR DESCRIPTION
**Pull request by Jacco**
*Issue #142 *

*Description of changes:*

- added "Condition" to the _supported_resource_attributes of Resource
- SamFunction now passes attributes to constructor of LambdaFunction in _construct_lambda_function with the accepted attribute values of the source
- added test case in test_translator.py (plus one input and three outputs) that checks "Condition" being created

This is my first PR, I know there is still work to do:

- the passing of the attributes should also work for other SamResourceMacro's
- The "Condition" attribute should also be put on other resources created on behalf of this one

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by Jacco** [comment]
Discussion:
Is the S3 push events a DependsOn is used. Since conditional DependsOn are not supported (AWS CloudFormation requires strings only), I used an undocumented way to introduce implicit dependencies. Its is descibed in this stackoverflow post:

[Conditional DependsOn](https://stackoverflow.com/questions/34607476/cloudformation-apply-condition-on-dependson)

In the code I add a Tag:

Key value: sam:ConditionalDependsOn:logical-id (logical-id of the permission in this case)
Value: Fn:If [TheCondition, Fn:GetAtt [logical-id, Arn], 'no dependency']

Regards Jacco

**Comment by keetonian** [comment]
@Jacco Thank you for submitting this PR! Are there any other outstanding items that you are going to address, or is this PR ready for review?

**Comment by brianium** [comment]
@brettstack is this blocked by those outstanding items mentioned by @Jacco ? Sounds like he is waiting on review? I know I'm new here, but is there anything I can do to help either of you?

**Comment by jlhood** [comment]
@Jacco Apologies for the delays in getting to this. We are still on the hook to review. @brianium, really appreciate the offer. Please feel free to take a look and comment on the PR if you see issues. Thanks!

File: samtranslator/model/eventsources/push.py [hunk 1]
```diff
@@ -253,6 +258,31 @@ def _depend_on_lambda_permissions(self, bucket, permission):
 
         return bucket
 
+    def _depend_on_lambda_permissions_using_tag(self, bucket, permission):
+        """
+        Since conditional DependsOn is not supported this undocumented way of
```
Comment (brettstack) [inline samtranslator/model/eventsources/push.py]: Add the explanation for this workaround (you can link to the StackOverflow but please also give an inline summary of how this works)

File: samtranslator/model/eventsources/push.py [hunk 2]
```diff
@@ -253,6 +258,31 @@ def _depend_on_lambda_permissions(self, bucket, permission):
...
+        tags = properties.get('Tags', None)
+        if tags is None:
+            tags = []
+            properties['Tags'] = tags
+        dep_tag = {
+            'sam:ConditionalDependsOn:' + permission.logical_id: {
+                'Fn:If': [
+                    permission.resource_attributes['Condition'],
+                    fnGetAtt(permission.logical_id, 'Arn'),
+                    'no dpendency'
```
Comment (brettstack) [inline samtranslator/model/eventsources/push.py]: typo

**Comment by keetonian** [comment]
@Jacco If you have a moment, could you take a look at #755? It's a continuation of your work here. Looking through it, the way to add a Conditional `depends_on` to an S3 bucket that you implemented is rather neat.

**Comment by Jacco** [comment]
@keetonian I will have a look. Am a bit busy ATM so it might take some time.

**Comment by keetonian** [comment]
Ok, no worries then. Thanks again for your work on this!

**Pull request merged**